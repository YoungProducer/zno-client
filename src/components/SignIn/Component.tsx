/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 28 February 2020
 *
 * Component which allows user to make sign in
 */

/** External imports */
import React, { useCallback, useState, useEffect } from 'react';
import classNames from 'classnames';
import Paper from '@material-ui/core/Paper';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import CheckBox from '@material-ui/core/Checkbox';
import InputBase from '@material-ui/core/InputBase';
import Typography from '@material-ui/core/Typography';
import EmailIcon from '@material-ui/icons/Email';
import LockIcon from '@material-ui/icons/Lock';
import { makeStyles, createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
import { useHistory } from 'react-router-dom';

/** Application's imports */
import { TSignInProps } from './container';
import NavigationLink from 'components/NavigationLink';
import Input from 'components/Input';
import Logo from 'img/logo';
import Check from 'public/images/check.svg';

/** Define classes as hook */
const useStyles = makeStyles((theme: Theme) => createStyles({
    backdrop: {
        background: '#fff',
        width: '100%',
        height: '100vh',
    },
    container: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        height: '100%',
    },
    root: {
        width: 'min-content',
        borderRadius: 4,
        filter: `drop-shadow(2.828px 2.828px 3.5px rgba(227,207,207,0.9))`,
        backgroundColor: '#ffffff',
        textAlign: 'center',
    },
    innerContainer: {
        borderRadius: 4,
        backgroundColor: '#f4f5f7',
        padding: theme.spacing(4),
    },
    logoContainer: {
        paddingTop: theme.spacing(1),
        paddingBottom: theme.spacing(1),
    },
    textField: {
        width: 230,
    },
    fields: {
        paddingBottom: theme.spacing(3),
    },
    actions: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    block: {
        width: 'max-content',
    },
    buttonBlock: {
        width: '100%',
        marginTop: theme.spacing(2),
    },
    button: {
        filter: `drop-shadow(0px 3px 2.5px rgba(0,0,0,0.16))`,
    },
    link: {
        color: theme.palette.primary.main,
    },
    icon: {
        color: '#555',
    },
    checkBoxIcon: {
        marginLeft: 2,
        borderRadius: 3,
        width: 20,
        height: 20,
        filter: `drop-shadow(2.828px 2.828px 3.5px rgba(227,207,207,0.9))`,
        backgroundColor: '#fff',
        backgroundImage: 'linear-gradient(180deg,hsla(0,0%,100%,.8),hsla(0,0%,100%,0))',
        '$root.Mui-focusVisible &': {
            outline: '2px auto rgba(19,124,189,.6)',
            outlineOffset: 2,
        },
        'input:hover ~ &': {
            backgroundColor: '#ebf1f5',
        },
        'input:disabled ~ &': {
            boxShadow: 'none',
            background: 'rgba(206,217,224,.5)',
        },
    },
    checkBoxСheckedIcon: {
        backgroundImage: 'linear-gradient(180deg,hsla(0,0%,100%,.8),hsla(0,0%,100%,0))',
        position: 'relative',
        '&:before': {
            position: 'absolute',
            top: 2,
            left: 2,
            backgroundImage: `url(${Check})`,
            width: 20,
            height: 20,
            backgroundSize: '16px, 16px',
            display: 'block',
            content: '""',
            backgroundRepeat: 'no-repeat',
        },
        'input:hover ~ &': {
            // backgroundColor: '#106ba3',
        },
    },
    checkBox: {
        width: 400,
    },
}));

/**
 * Create hook to make component stateless.
 * Compute all values from text fields.
 * Returns fields data and button onClick property.
 */
const useSignInElements = (props: TSignInProps) => {
    const {
        isLoggedIn,
        errorFields,
        fieldsMessages,
        fetchSignIn,
        setSignInErrorFieldsToDefault,
        setSignInFieldsMessagesToDefault,
    } = props;

    const history = useHistory();

    useEffect(() => {
        if (isLoggedIn) {
            const next = history.location.state
                ? `${(history.location.state as any).from.pathname}${(history.location.state as any).from.search}`
                : undefined;
            // const next: string = new URLSearchParams(history.location.state.).get('next');
            history.push(next || '/');
        }
    }, [isLoggedIn]);

    const clearFields = () => {
        setSignInErrorFieldsToDefault();
        setSignInFieldsMessagesToDefault();
    };

    /** User email */
    const [email, setEmail] = useState<string>('');
    const handleChangeEmail =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setEmail(event.target.value);
            if (errorFields.email) {
                clearFields();
            }
        }, [errorFields.email]);

    /** User password */
    const [password, setPassword] = useState<string>('');
    const handleChangePassword =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setPassword(event.target.value);
            if (errorFields.password) {
                clearFields();
            }
        }, [errorFields.password]);

    /** SignUp button handle */
    const handleFetchSignIn = useCallback(() => fetchSignIn({
        email,
        password,
    }), [email, password]);

    return {
        emailField: {
            onChange: handleChangeEmail,
            value: email,
            error: errorFields.email,
            helperText: fieldsMessages.email,
        },
        passwordField: {
            onChange: handleChangePassword,
            value: password,
            error: errorFields.password,
            helperText: fieldsMessages.password,
        },
        signInButton: {
            onClick: handleFetchSignIn,
        },
    };
};

/** Create component */
const Component = (props: TSignInProps) => {
    /** Create styles */
    const classes = useStyles({});

    /** Get fields data from hook */
    const { emailField, passwordField, signInButton } = useSignInElements(props);

    return (
        <div
            className='auth-background'
        >
            <Container maxWidth='lg' className={classes.container}>
                <Paper
                    className={classes.root}
                    elevation={0}
                    square
                >
                    <div className={classes.logoContainer}>
                        <Logo />
                    </div>
                    <div className={classes.innerContainer}>
                        <Grid
                            container
                            direction='column'
                            alignItems='center'
                            justify='flex-start'
                            item
                            spacing={3}
                            className={classes.block}
                        >
                            <Grid item>
                                <Input
                                    color='primary'
                                    type='email'
                                    placeholder='Емеїл'
                                    startAdornment={<EmailIcon className={classes.icon}/>}
                                    data-testid='signin-email-input'
                                    {...emailField}
                                />
                            </Grid>
                            <Grid item>
                                <Input
                                    color='primary'
                                    type='password'
                                    placeholder='Пароль'
                                    startAdornment={<LockIcon className={classes.icon}/>}
                                    data-testid='signin-password-input'
                                    {...passwordField}
                                />
                            </Grid>
                            <Grid item>
                                <FormControl component='fieldset' className={classes.checkBox}>
                                    <FormControlLabel
                                        control={
                                            <CheckBox
                                                disableRipple
                                                disableTouchRipple
                                                disableFocusRipple
                                                color='primary'
                                                checkedIcon={<span className={classNames(classes.checkBoxIcon, classes.checkBoxСheckedIcon)} />}
                                                icon={<span className={classes.checkBoxIcon}/>}
                                            />
                                        }
                                        label={`Запам'ятати мене`}
                                    />
                                </FormControl>
                            </Grid>
                            <Grid item>
                                <Button
                                    color='primary'
                                    variant='contained'
                                    disableElevation
                                    className={classes.button}
                                    data-testid='signin-button'
                                    { ...signInButton }
                                >
                                    Увійти
                                </Button>
                            </Grid>
                            <Grid item>
                                <Typography align='center'>
                                    Не зареєстровані?
                                    <NavigationLink
                                        navLink={{
                                            to: '/auth/signup',
                                        }}
                                        className={classes.link}
                                    >
                                        Зареєструватися
                                    </NavigationLink>
                                </Typography>
                            </Grid>
                        </Grid>
                    </div>
                </Paper>
            </Container>
        </div>
    );
};

export default Component;
