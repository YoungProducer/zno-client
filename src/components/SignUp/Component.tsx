/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 25 February 2020
 *
 * Component which allows create user's account.
 */

// External imports
import React, { useState, useCallback, useEffect, memo, useMemo } from 'react';
import Paper from '@material-ui/core/Paper';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

// Application's imports
import { TSignUpProps } from './container';
import NavigationLink from 'components/NavigationLink';

// Define classes as hook
const useStyles = makeStyles((theme: Theme) => createStyles({
    backdrop: {
        background: '#fff',
        width: '100%',
        height: '100vh',
    },
    root: {
        position: 'fixed',
        padding: 15,
        left: '50%',
        top: '50%',
        transform: `translate(-50%, -50%)`,
        width: 'min-content',
    },
    textField: {
        width: 230,
    },
    fields: {
        paddingBottom: theme.spacing(3),
    },
    actions: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    block: {
        width: 'max-content',
    },
    button: {
        width: 230,
    },
    link: {
        color: theme.palette.primary.main,
    },
}));

/**
 * Create hook to make component stateless.
 * Compute all values from text fields.
 * Returns fields data and button onClick property.
 */
const useSignUpElements = (props: TSignUpProps) => {
    const {
        errorFields,
        fieldsMessages,
        fetchSignUp,
        setSignUpErrorFieldsToDefault,
        setSignUpFieldsMessagesToDefault,
    } = props;

    const clearFields = () => {
        setSignUpErrorFieldsToDefault();
        setSignUpFieldsMessagesToDefault();
    };

    /** User email */
    const [email, setEmail] = useState<string>('');
    const handleChangeEmail =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setEmail(event.target.value);
            if (errorFields.email) {
                clearFields();
            }
        }, [errorFields.email]);

    /** User password */
    const [password, setPassword] = useState<string>('');
    const handleChangePassword =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setPassword(event.target.value);
            if (errorFields.password) {
                clearFields();
            }
        }, [errorFields.password]);

    /** Confirmation password, should be the same to the password */
    const [confPassword, setConfPassword] = useState<string>('');
    const handleChangeConfPassword =
        useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
            setConfPassword(event.target.value);
            if (errorFields.confPassword) {
                clearFields();
            }
        }, [errorFields.confPassword]);

    /** SignUp button handle */
    const handleFetchSignUp = useCallback(() => fetchSignUp({
        email,
        password,
        confPassword,
    }), [email, password, confPassword]);

    return {
        emailField: {
            onChange: handleChangeEmail,
            value: email,
            error: errorFields.email,
            helperText: fieldsMessages.email,
        },
        passwordField: {
            onChange: handleChangePassword,
            value: password,
            error: errorFields.password,
            helperText: fieldsMessages.password,
        },
        confPasswordField: {
            onChange: handleChangeConfPassword,
            value: confPassword,
            error: errorFields.confPassword,
            helperText: fieldsMessages.confPassword,
        },
        signUpButton: {
            onClick: handleFetchSignUp,
        },
    };
};

const Component = (props: TSignUpProps) => {
    /** Create object with classes to use them */
    const classes = useStyles({});

    /** Get fields data from hook */
    const {
        emailField,
        passwordField,
        confPasswordField,
        signUpButton,
    } = useSignUpElements(props);

    return (
        <div className={classes.backdrop}>
            <Paper
                className={classes.root}
                elevation={2}
                square
            >
                <Typography
                    color='primary'
                    variant='h6'
                    align='center'
                >
                    Реєстрація
                </Typography>
                <Grid
                    container
                    direction='row'
                    spacing={2}
                    className={classes.fields}
                >
                    <Grid
                        container
                        direction='row'
                        alignItems='center'
                        item
                        spacing={3}
                        className={classes.block}
                    >
                        <Grid item>
                            <TextField
                                className={classes.textField}
                                color='primary'
                                label='Емеїл'
                                type='email'
                                { ...emailField }
                            />
                        </Grid>
                        <Grid item>
                            <TextField
                                className={classes.textField}
                                color='primary'
                                label='Пароль'
                                type='password'
                                { ...passwordField }
                            />
                        </Grid>
                    </Grid>
                    <Grid
                        container
                        direction='row'
                        alignItems='center'
                        item
                        spacing={3}
                        className={classes.block}
                    >
                        <Grid item>
                            <TextField
                                className={classes.textField}
                                color='primary'
                                label='Пароль підтвердження'
                                type='password'
                                { ...confPasswordField }
                                { ...{ 'data-testid': 'conf-password' } }
                            />
                        </Grid>
                        <Grid item>
                            <Button
                                color='primary'
                                variant='contained'
                                disableElevation
                                className={classes.button}
                                data-testid='signup-button'
                                { ...signUpButton }
                            >
                                Зареєструватися
                            </Button>
                        </Grid>
                    </Grid>
                </Grid>
                <Typography align='center'>
                    Вже зареєстровані?
                    <NavigationLink
                        navLink={{
                            to: '/auth/signin',
                        }}
                        className={classes.link}
                    >
                        Увійти
                    </NavigationLink>
                </Typography>
            </Paper>
        </div>
    );
};

export default Component;
