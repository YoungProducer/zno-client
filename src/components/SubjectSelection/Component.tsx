/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 1 March 2020
 *
 * Component to select subjects.
 */

/** External imports */
import React, { useEffect } from 'react';
import classNames from 'classnames';
import { useHistory } from 'react-router-dom';
import Container from '@material-ui/core/Container';
import InputBase from '@material-ui/core/InputBase';
import IconButton from '@material-ui/core/IconButton';
import SearchIcon from '@material-ui/icons/Search';

/** Application's imports */
import { TSubjectSelectionProps } from './container';
import SubjectTile from './SubjectTile';
import SubjectPresentation from './SubjectsPresentation';
import { makeStyles, Theme, createStyles } from '@material-ui/core';

const useStyles = makeStyles((theme: Theme) => createStyles({
    root: {
        width: '100%',
        height: '100vh',
    },
    container: {
        display: 'flex',
    },
    searchBlock: {
        width: '50%',
    },
    popularBlock: {
        background: '#5c498c',
        height: 85,
        width: 'max-content',
        position: 'absolute',
        // bottom: '1%',
        bottom: '1%',
        right: 0,
        left: 0,
        margin: 'auto',
        borderRadius: 7,
        paddingLeft: theme.spacing(3),
        paddingRight: theme.spacing(3),
    },
    popularBlockInner: {
        height: 'inherit',
        width: '100%',
    },
    searchInput: {
        width: '90%',
        height: 67,
        background: '#4d8fcb',
        borderRadius: 34,
        paddingLeft: theme.spacing(2),
        paddingRight: 6,
        color: '#fff',
        fontSize: '1.3rem',
    },
    iconButton: {
        width: 55,
        height: 55,
        background: '#fff',
        borderRadius: '50%',
        '&:hover': {
            background: '#f1f1f1',
        },
    },
    icon: {
        color: '#5955cf',
        width: 30,
        height: 30,
    },
}));

/** Create component */
const Component = (props: TSubjectSelectionProps) => {
    const classes = useStyles({});

    const {
        // classes,
        subjectsList,
        fetchSubjectsNames,
    } = props;

    useEffect(() => {
        fetchSubjectsNames();
    }, []);

    return (
        <>
            <div className={classNames(classes.root, 'subject-selection-background')}>
                <Container maxWidth='lg' className={classes.container}>
                    <div className={classes.searchBlock}>
                        <InputBase
                            className={classes.searchInput}
                            placeholder='Знайти предмет'
                            endAdornment={
                                <IconButton
                                    size='small'
                                    className={classes.iconButton}
                                >
                                    <SearchIcon
                                        className={classes.icon}
                                    />
                                </IconButton>
                            }
                        />
                    </div>
                    <SubjectPresentation />
                </Container>
            </div>
            {/* <div className={classes.popularBlock}>
                <SubjectTile
                    subject={subjectsList[0]}
                />
                <SubjectTile
                    subject={subjectsList[1]}
                />
            </div> */}
        </>
    );
};

export default Component;
