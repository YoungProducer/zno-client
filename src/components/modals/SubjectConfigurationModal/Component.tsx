/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 4 March 2020
 *
 * Component which allows to configure test options
 * and allows to select which test you want to complete.
 */

/** External imports */
import React, { useState, useCallback, useMemo, useEffect } from 'react';
import Dialog from '@material-ui/core/Dialog';
import MuiDialogTitle from '@material-ui/core/DialogTitle';
import MuiDialogContent from '@material-ui/core/DialogContent';
import MuiDialogActions from '@material-ui/core/DialogActions';
import Typography from '@material-ui/core/Typography';
import MenuItem from '@material-ui/core/MenuItem';
import TextField from '@material-ui/core/TextField';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormControl from '@material-ui/core/FormControl';
import FormLabel from '@material-ui/core/FormLabel';
import Button from '@material-ui/core/Button';
import { createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';

/** Application's imports */
import {
    TTestTypes,
    ETestTypes,
    TSubjectConfigurationModalProps,
} from './container';

/** Define Material UI styles as hook */

const styles = (theme: Theme) =>
    createStyles({
        root: {
            margin: 0,
            padding: theme.spacing(2),
        },
    });

/** Create components wrapped by withStyles */
export interface DialogTitleProps extends WithStyles<typeof styles> {
    id: string;
    children: React.ReactNode;
}

const DialogTitle = withStyles(styles)((props: DialogTitleProps) => {
    const { children, classes, ...other } = props;
    return (
        <MuiDialogTitle disableTypography className={classes.root} {...other}>
            <Typography variant="h6">{children}</Typography>
        </MuiDialogTitle>
    );
});

const DialogContent = withStyles((theme: Theme) => ({
    root: {
        padding: theme.spacing(2),
        paddingLeft: theme.spacing(3),
        paddingRight: theme.spacing(3),
    },
}))(MuiDialogContent);

/**
 * Define custom hook to save state of configuration.
 * This hook returns object with settings
 * for all input elements and dialog.
 */
const useSubjectConfigurationElements = (props: TSubjectConfigurationModalProps) => {
    /** Destruct props */
    const {
        dialogVisible,
        subjectThemes,
        subSubjectsNames,
        subSubjectsThemes,
        toggleSubjectConfigurationDialog,
    } = props;

    /** Responsible for themes or exams selection */
    const [testType, setTestType] = useState<ETestTypes>('' as ETestTypes);
    const handleChangeTestType = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) =>
            setTestType(event.target.value as ETestTypes),
        []);

    /** Responsible for theme selection of specific sub-subject */
    const [subSubject, setSubSubject] = useState<string>(subSubjectsNames !== null ? subSubjectsNames[0] : '');
    const handleChageSubSubject = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) =>
            setSubSubject(event.target.value),
        []);

    /** Responsible for theme selection */
    const [theme, setTheme] = useState<string>(subjectThemes !== null ? subjectThemes[0] : '');
    const handleChangeTheme = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) =>
            setTheme(event.target.value),
        []);

    /** Memoized value which returns list of sub-subjects */
    const subSubjects = useMemo(() => {
        if (subSubjectsNames === null || testType !== ETestTypes.THEMES) return null;

        return subSubjectsNames.map(subject => (
            <MenuItem key={subject} value={subject}>
                {subject}
            </MenuItem>
        ));
    }, [subSubjectsNames, testType]);

    /** Responsible for displaying of sub-subject selection */
    const displaySubSubjectSelection = useMemo(() =>
        subSubjectsNames !== null && testType === ETestTypes.THEMES,
    [testType, subSubjectsNames]);

    /** Memoized value which returns list of themes */
    const themes = useMemo(() => {
        /**
         * If subSubjectsThemes is not null and some subSubject is selected
         * and testType equals 'THEMES'
         * then this function will returns themes list for selected subSubject.
         */
        if (subSubjectsThemes !== null && subSubject !== '' && testType === ETestTypes.THEMES) {
            return subSubjectsThemes[subSubject].map(theme => (
                <MenuItem key={theme} value={theme}>
                    {theme}
                </MenuItem>
            )) || null;
        }

        /**
         * If subjectThemes is not null and subSubjectsThemes is null
         * and testType equals 'THEMES'
         * then this function will returns themes list for current subject.
         */
        if (subjectThemes !== null && subSubjectsThemes === null && testType === ETestTypes.THEMES) {
            return subjectThemes.map(theme => (
                <MenuItem key={theme} value={theme}>
                    {theme}
                </MenuItem>
            ));
        }

        return null;
    }, [subjectThemes, testType, subSubject, subSubjectsThemes]);

    /** Responsible for displaying of theme selection */
    const displayThemeSelection = useMemo(() =>
        subjectThemes !== null && testType === ETestTypes.THEMES,
    [testType, subjectThemes]);

    return {
        displayThemeSelection,
        displaySubSubjectSelection,
        subSubjects,
        themes,
        dialog: {
            open: dialogVisible,
            onClose: () => toggleSubjectConfigurationDialog(false),
        },
        selectTypeField: {
            value: testType,
            onChange: handleChangeTestType,
        },
        selectSubSubjectField: {
            value: subSubject,
            onChange: handleChageSubSubject,
        },
        selectThemeField: {
            value: theme,
            onChange: handleChangeTheme,
        },
    };
};

/** Create component */
const Component = (props: TSubjectConfigurationModalProps) => {
    /** Get elements from custom hook */
    const {
        dialog,
        themes,
        subSubjects,
        selectTypeField,
        selectThemeField,
        selectSubSubjectField,
        displayThemeSelection,
        displaySubSubjectSelection,
    } = useSubjectConfigurationElements(props);

    return (
        <Dialog
            {...dialog}
            aria-labelledby='subject-config-title'
            aria-describedby=''
        >
            <DialogTitle id='subject-config-title'>
                Налаштування тесту
            </DialogTitle>
            <DialogContent>
                <FormControl component='fieldset'>
                    <FormLabel component='legend'>Оберіть тип тесту</FormLabel>
                    <RadioGroup
                        aria-label='select-test-type'
                        name='select-test-type'
                        data-testid='select-test-type'
                        {...selectTypeField}
                    >
                        <FormControlLabel
                            value={ETestTypes.THEMES}
                            control={<Radio color='primary' />}
                            label={ETestTypes.THEMES}
                            labelPlacement='end'
                        />
                        <FormControlLabel
                            value={ETestTypes.EXAMS}
                            control={<Radio color='primary' />}
                            label={ETestTypes.EXAMS}
                            labelPlacement='end'
                        />
                    </RadioGroup>
                </FormControl>

                { displaySubSubjectSelection && (
                    <FormControl component='div'>
                        <FormLabel component='legend'>Оберіть предмет</FormLabel>
                        <TextField
                            id='standard-select-currency'
                            select={true}
                            {...selectSubSubjectField}
                            margin='none'
                            variant='outlined'
                            data-testid='select-sub-subject'
                        >
                            {subSubjects}
                        </TextField>
                    </FormControl>
                )}

                { displayThemeSelection && (
                    <FormControl component='div'>
                        <FormLabel component='legend'>Оберіть тему</FormLabel>
                        <TextField
                            id='standard-select-currency'
                            select={true}
                            {...selectThemeField}
                            margin='none'
                            variant='outlined'
                            data-testid='select-subject-theme'
                        >
                            {themes}
                        </TextField>
                    </FormControl>
                )}
            </DialogContent>
        </Dialog>
    );
};

/** Export component */
export default Component;
