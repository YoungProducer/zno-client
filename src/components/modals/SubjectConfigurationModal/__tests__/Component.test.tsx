/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 4 March 2020
 *
 * Create test suites for SubjectConfigurationModal component.
 * @jest-environment jsdom
 */

/** External imports */
import React from 'react';
import { shallow, mount } from 'enzyme';
import Backdrop from '@material-ui/core/Backdrop';
import MenuItem from '@material-ui/core/MenuItem';

/** Application's imports */
import Component from '../Component';
import { ETestTypes, TSubjectConfigurationModalProps } from '../container';

describe('SubjectConfigurationModal component', () => {
    /** Create mocked functions for props */
    const fetchSubjectConfiguration = jest.fn();
    const toggleSubjectConfigurationDialog = jest.fn();

    /** Define required props */
    const requiredProps: TSubjectConfigurationModalProps = {
        fetchSubjectConfiguration,
        toggleSubjectConfigurationDialog,
        dialogVisible: true,
        loading: false,
        subjectName: 'foo',
        subSubjectsNames: null,
        subSubjectsThemes: null,
        subjectExams: null,
        subjectThemes: null,
    };

    beforeEach(() => {
        (requiredProps.fetchSubjectConfiguration as jest.Mock).mockReset();
        (requiredProps.toggleSubjectConfigurationDialog as jest.Mock).mockReset();
    });

    test('Is match snapshot', () => {
        /** Render component in shallow via enzyme */
        const tree = shallow(<Component {...requiredProps} />);

        /** Assert component matches snapshot */
        expect(tree).toMatchSnapshot();
    });

    test('Select test type', () => {
        /** Render component */
        const tree = shallow(<Component {...requiredProps} />);

        /** Simulate change event */
        tree.find(`[data-testid='select-test-type']`).simulate('change', {
            target: {
                value: ETestTypes.THEMES,
            },
        });

        /** Assert testType has correct value */
        expect(tree.find(`[data-testid='select-test-type']`).props().value).toEqual(ETestTypes.THEMES);
    });

    test('Dialog onClose handler', () => {
        /** Mount component */
        const tree = mount(<Component {...requiredProps} />);

        /** Simulate click outside of dialog */
        tree.find(Backdrop).simulate('click');

        /** Assert toggleSubjectConfigurationDialog have been called with right arg  */
        expect(toggleSubjectConfigurationDialog).toHaveBeenCalledWith(false);
    });

    test(`If subSubjectsNames is not null and testType equals 'THEMES' component should render sub-subject selection text-field`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsNames={['foo']}
            />,
        );

        /** Assert sub-subject selection text-field doesn't exist */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeFalsy();

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert sub-subject selection text-field exists */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeTruthy();
    });

    test(`If subSubjectsNames is null select-sub-subject text field shouldn't exists`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsNames={null}
            />,
        );

        /** Assert sub-subject selection text-field doesn't exist */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeFalsy();

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert sub-subject selection text-field exists */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeFalsy();
    });

    test(`If subSubjectsNames is not null and testType equals 'THEMES' select sub-subject`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsNames={['foo']}
            />,
        );

        /** Assert sub-subject selection text-field doesn't exist */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeFalsy();

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert sub-subject selection text-field exists */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeTruthy();

        /** Simulate onChange event in select-sub-subject text-field */
        tree.find(`[data-testid='select-sub-subject']`).simulate('change', { target: { value: 'foo' } });

        /** Assert select-sub-subject has value 'foo' */
        expect(tree.find(`[data-testid='select-sub-subject']`).props().value).toEqual('foo');
    });

    test(`If subjectThemes is null select-subject-theme text field shouldn't exist`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={null}
                subSubjectsNames={null}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-sub-themes selection doesn't exist */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeFalsy();
    });

    test(`If subjectThemes is not null and testType equals 'Themes' select-subject-theme text field should exists`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={['bar']}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-sub-themes selection exists */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeTruthy();
    });

    test(`If subjectThemes is not null and testType equals 'Themes' select theme`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={['bar']}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme selection text-field exists */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeTruthy();

        /** Simulate onChange event in select-subject-theme text field */
        tree.find(`[data-testid='select-subject-theme']`).simulate('change', { target: { value: 'bar' } });

        /** Assert select-subject-theme text field has correct value */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('bar');
    });

    test(`If subSubjectThemes is not null and testType equals 'Themes' component should display theme selection`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsNames={['foo']}
                subSubjectsThemes={{ foo: ['bar'] }}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme selection text-field exists */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeTruthy();
    });

    test(`If subSubjectName doesn't exist in subSubjectThemes`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsNames={['foo']}
                subSubjectsThemes={{ zoo: ['bar'] }}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme selection text-field exists */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeTruthy();
    });
});
