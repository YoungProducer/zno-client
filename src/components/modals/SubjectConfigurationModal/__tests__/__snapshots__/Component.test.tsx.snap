// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubjectConfigurationModal component Is match snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-describedby=""
  aria-labelledby="subject-config-title"
  onClose={[Function]}
  open={true}
>
  <WithStyles(Component)
    id="subject-config-title"
  >
    Налаштування тесту
  </WithStyles(Component)>
  <WithStyles(WithStyles(ForwardRef(DialogContent)))>
    <WithStyles(ForwardRef(FormControl))
      component="fieldset"
    >
      <WithStyles(ForwardRef(FormLabel))
        component="legend"
      >
        Оберіть тип тесту
      </WithStyles(ForwardRef(FormLabel))>
      <ForwardRef(RadioGroup)
        aria-label="select-test-type"
        data-testid="select-test-type"
        name="select-test-type"
        onChange={[Function]}
        value=""
      >
        <WithStyles(ForwardRef(FormControlLabel))
          control={
            <WithStyles(ForwardRef(Radio))
              color="primary"
            />
          }
          label="Вибір теми"
          labelPlacement="end"
          value="Вибір теми"
        />
        <WithStyles(ForwardRef(FormControlLabel))
          control={
            <WithStyles(ForwardRef(Radio))
              color="primary"
            />
          }
          label="Тренувальні варіанти ЗНО"
          labelPlacement="end"
          value="Тренувальні варіанти ЗНО"
        />
      </ForwardRef(RadioGroup)>
    </WithStyles(ForwardRef(FormControl))>
  </WithStyles(WithStyles(ForwardRef(DialogContent)))>
</WithStyles(ForwardRef(Dialog))>
`;
